### PR TITLE
Fix indexing with datetime64[ns] with pandas=1.1

### LINKED ISSUE #4283: Selection with datetime64[ns] fails with Pandas 1.1.0
I ran into this issue with a netCDF file with the following time variable:
```
	double time1(time1) ;
		time1:_FillValue = NaN ;
		time1:standard_name = "time" ;
		time1:long_name = "time" ;
		time1:udunits = "Hour since 2017-09-05T12:00:00Z" ;
		time1:units = "Hour since 2017-09-05T12:00:00+00:00" ;
		time1:calendar = "proleptic_gregorian" ;

 time1 = 0, 3, 6, 9, 12, 15, 18, 21, 24 ;
```
but we can reproduce the problem with something as simple as:
```python
import numpy as np
import xarray as xr

t = np.array(['2017-09-05T12:00:00.000000000', '2017-09-05T15:00:00.000000000'], dtype='datetime64[ns]')
da = xr.DataArray(np.ones(t.shape), dims=('time',), coords=(t,))

da.loc[{'time':t[0]}]  # Works on pandas 1.0.5
```
this produces:
```pytb
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-11-3e0afa0bd195> in <module>
----> 1 da.loc[{'time':t[0]}]

~/miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/dataarray.py in __getitem__(self, key)
    196             labels = indexing.expanded_indexer(key, self.data_array.ndim)
    197             key = dict(zip(self.data_array.dims, labels))
--> 198         return self.data_array.sel(**key)
    199 
    200     def __setitem__(self, key, value) -> None:

~/miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/dataarray.py in sel(self, indexers, method, tolerance, drop, **indexers_kwargs)
   1147 
   1148         """
-> 1149         ds = self._to_temp_dataset().sel(
   1150             indexers=indexers,
   1151             drop=drop,

~/miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/dataset.py in sel(self, indexers, method, tolerance, drop, **indexers_kwargs)
   2099         """
   2100         indexers = either_dict_or_kwargs(indexers, indexers_kwargs, "sel")
-> 2101         pos_indexers, new_indexes = remap_label_indexers(
   2102             self, indexers=indexers, method=method, tolerance=tolerance
   2103         )

~/miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/coordinates.py in remap_label_indexers(obj, indexers, method, tolerance, **indexers_kwargs)
    394     }
    395 
--> 396     pos_indexers, new_indexes = indexing.remap_label_indexers(
    397         obj, v_indexers, method=method, tolerance=tolerance
    398     )

~/miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/indexing.py in remap_label_indexers(data_obj, indexers, method, tolerance)
    268             coords_dtype = data_obj.coords[dim].dtype
    269             label = maybe_cast_to_coords_dtype(label, coords_dtype)
--> 270             idxr, new_idx = convert_label_indexer(index, label, dim, method, tolerance)
    271             pos_indexers[dim] = idxr
    272             if new_idx is not None:

~/miniconda3/envs/py38/lib/python3.8/site-packages/xarray/core/indexing.py in convert_label_indexer(index, label, index_name, method, tolerance)
    187                 indexer = index.get_loc(label.item())
    188             else:
--> 189                 indexer = index.get_loc(
    190                     label.item(), method=method, tolerance=tolerance
    191                 )

~/miniconda3/envs/py38/lib/python3.8/site-packages/pandas/core/indexes/datetimes.py in get_loc(self, key, method, tolerance)
    620         else:
    621             # unrecognized type
--> 622             raise KeyError(key)
    623 
    624         try:

KeyError: 1504612800000000000
```

what's interesting is changing the units of `datetime64` to `[s]` works:
```python
import numpy as np
import xarray as xr

t = np.array(['2017-09-05T12:00:00.000000000', '2017-09-05T15:00:00.000000000'], dtype='datetime64[s]')
da = xr.DataArray(np.ones(t.shape), dims=('time',), coords=(t,))
da.loc[{'time':t[0]}]  # Works
```

**Environment**:
Python 3.8 from conda-forge on macOS 10.15.4

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.8.5 | packaged by conda-forge | (default, Jul 24 2020, 01:06:20) 
[Clang 10.0.1 ]
python-bits: 64
OS: Darwin
OS-release: 19.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.16.0
pandas: 1.1.0
numpy: 1.19.1
scipy: 1.5.2
netCDF4: 1.5.4
pydap: None
h5netcdf: None
h5py: 2.10.0
Nio: None
zarr: None
cftime: 1.2.1
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: 0.9.8.3
iris: None
bottleneck: None
dask: 2.21.0
distributed: 2.21.0
matplotlib: 3.3.0
cartopy: 0.18.0
seaborn: None
numbagg: None
pint: 0.14
setuptools: 49.2.0.post20200712
pip: 20.1.1
conda: None
pytest: 6.0.0
IPython: 7.16.1
sphinx: 2.4.4

</details>

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #4283

The underlying issue is that calling `.item()` on a NumPy array with
`dtype=datetime64[ns]` returns an _integer_, rather than an `np.datetime64`
scalar. This is somewhat baffling but works this way because `.item()`
returns native Python types, but `datetime.datetime` doesn't support
nanosecond precision.

`pandas.Index.get_loc` used to support these integers, but now is more strict.
Hence we get errors.

We can fix this by using `array[()]` to convert 0d arrays into NumPy scalars
instead of calling `array.item()`.

I've added a crude regression test. There may well be a better way to test this
but I haven't figured it out yet.

 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`

### CODE REVIEW

**Comment by keewis** [comment]
you should be able to isolate this to just `indexing.convert_label_indexer`, e.g. with
```python
    def test_convert_label_indexer_datetime(self):
        index = pd.to_datetime(["2000-01-01", "2001-01-01", "2002-01-01"])
        actual = indexing.convert_label_indexer(index, "2001-01-01")
        expected = (1, None)
        assert actual == expected

        actual = indexing.convert_label_indexer(index, index.to_numpy()[1])
        assert actual == expected
```

The failing tests are due to `label[()]` returning a `numpy.str_` instead of a plain python `str`. Maybe we can fix that by using `item` as long as the `dtype` is not "datetime64" or "timedelta64":
```python
if label.dtype.kind in "mM":
    label_value = label[()]
else:
    label_value = label.item()
```

File: xarray/core/indexing.py [hunk 1]
```diff
@@ -173,8 +173,9 @@ def convert_label_indexer(index, label, index_name="", method=None, tolerance=No
             else _asarray_tuplesafe(label)
```
Comment (shoyer) [inline xarray/core/indexing.py:173]: Maybe we could avoid casting to ndarray on this line? All we really need it for is checking dimensionality of the result.

Comment (keewis) [reply xarray/core/indexing.py:173]: we're also using `label.dtype.kind`, so I'm not sure how we can avoid casting to ndarray?

Comment (keewis) [reply xarray/core/indexing.py:173]: I guess we could simply extract `label.ndim` and `label.dtype`?

**Comment by keewis** [comment]
there are lots of people that stumble into this, so I think it might be good to get this to work as soon as possible and issue a bugfix release.

**Comment by keewis** [comment]
gentle ping, @shoyer. Any updates on this?

**Comment by keewis** [comment]
it seems we *have* to cast because `label` may also be something like a `Variable` object, which is not accepted by `index.get_loc`. I pushed my original fix (falling back to `.item()` for non-datetime / timedelta dtypes), I hope that's okay?

**Comment by keewis** [comment]
it seems `pandas` warns about our usage of `pandas.Grouper`:
```
/home/docs/checkouts/readthedocs.org/user_builds/xray/checkouts/4292/xarray/core/common.py:1134: FutureWarning: 'base' in .resample() and in Grouper() is deprecated.
The new arguments that you should use are 'offset' or 'origin'.

>>> df.resample(freq="3s", base=2)

becomes:

>>> df.resample(freq="3s", offset="2s")

  grouper = pd.Grouper(
```
this was introduced in `1.1.0`. We're supporting `pandas>=0.25` (maybe even `>=0.24`) so we can't switch yet. I added a warning filter and a todo comment, but we might also need a tracking issue.

File: xarray/core/indexing.py [hunk 2]
```diff
@@ -178,8 +178,9 @@ def convert_label_indexer(index, label, index_name="", method=None, tolerance=No
...
         )
         if label.ndim == 0:
+            label_value = label[()] if label.dtype.kind in "mM" else label.item()
```
Comment (max-sixty) [inline xarray/core/indexing.py:181]:
```suggestion
            # see https://github.com/pydata/xarray/pull/4292/files for details
            label_value = label[()] if label.dtype.kind in "mM" else label.item()
```

Comment (keewis) [reply xarray/core/indexing.py:181]: is there a reason why you point to the changes of this PR? If not, I would use
```suggestion
            # see https://github.com/pydata/xarray/pull/4292 for details
            label_value = label[()] if label.dtype.kind in "mM" else label.item()
```

**Comment by dcherian** [comment]
I am +1 on merging this quickly and issuing a bugfix release. 

We can always make cleanups later...

File: xarray/core/indexing.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/indexing.py:181]: Agree. Will commit yours for the sake of expediancy

**Comment by shoyer** [comment]
OK, submitting this!

Thanks @keewis for making this fix actually work :)

**Comment by seth-p** [comment]
Does this fix #4363?

**Comment by shoyer** [comment]
> Does this fix #4363?

No, that seems to be unrelated

**Comment by max-sixty** [comment]
I can do a patch release this weekend

**Pull request merged**